Source/WebCore/page/Page.h

```diff
@@ -1390,8 +1390,31 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
     void updateDisplayEDRSuppression();
 #endif
 
-    // Checking hardware keyboard attached
-    void setHardwareKeyboardAttached(bool attached) { m_hardwareKeyboardAttached = attached; }
+    using HardwareKeyboardAttachmentObserver = Function<void(bool)>;
+    Vector<HardwareKeyboardAttachmentObserver> m_hardwareKeyboardAttachmentObservers;
+
+    void addHardwareKeyboardAttachmentObserver(HardwareKeyboardAttachmentObserver&& observer)
+    {
+        m_hardwareKeyboardAttachmentObservers.append(WTFMove(observer));
+    }
+
+    void flushHardwareKeyboardAttachmentObservers()
+    {
+        bool attached = m_hardwareKeyboardAttached;
+        std::ranges::for_each(m_hardwareKeyboardAttachmentObservers, [attached](auto& observer) {
+            observer(attached);
+        });
+        m_hardwareKeyboardAttachmentObservers.clear();
+    }
+
+    void didUpdateHardwareKeyboardAttachment(bool attached)
+    {
+        if (m_hardwareKeyboardAttached == attached)
+            return;
+
+        m_hardwareKeyboardAttached = attached;
+    }
```

Comment:
I would expect `flushHardwareKeyboardAttachmentObservers()` to be an implementation detail private to `WebCore::Page` that gets called automatically whenever we alter `m_hardwareKeyboardAttached` in `didUpdateHardwareKeyboardAttachment()`.

It seems odd that `WebPage` has the duty to call it, and even then we don't call it unconditionally?

Reply:
More concretely, I would expect:

```cpp
    void didUpdateHardwareKeyboardAttachment(bool attached)
    {
        if (m_hardwareKeyboardAttached == attached)
            return;

        m_hardwareKeyboardAttached = attached;
        flushHardwareKeyboardAttachmentObservers();
    }
```

... plus, that lets us make `flushHardwareKeyboardAttachmentObservers()` a private method.